File: 113-path-sum-ii/113-path-sum-ii.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    void helper(int target, TreeNode* root, vector<int> temp, vector<vector<int>>& res){
        if(!root)return;
        temp.push_back(root->val);
        target-=root->val;
        if(target==0 && !root->left && !root->right){
            res.push_back(temp);
            return;
        }
        helper(target,root->left,temp,res);
        helper(target,root->right,temp,res);
    }
    vector<vector<int>> pathSum(TreeNode* root, int targetSum) {
        vector<vector<int>> res;
        if(!root)return res;
        helper(targetSum,root,vector<int>(),res);
        return res;
    }
};

File: 113-path-sum-ii/113-path-sum-ii.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    List<List<Integer>> res = new ArrayList<>();
    
    void helper(TreeNode root, int sum, int target, List<Integer> path){
        if(root==null)return;
        if(root.left==null && root.right==null && sum==target){
            res.add(new ArrayList(path));
            return;
        }
        if(root.left!=null){
            path.add(root.left.val);
            helper(root.left, sum+root.left.val,target,path);
            int index = path.size()-1;
            path.remove(index);
        }
        if(root.right!=null){
            path.add(root.right.val);
            helper(root.right, sum+root.right.val,target,path);
            int index = path.size()-1;
            path.remove(index);
        }
    }
    
    public List<List<Integer>> pathSum(TreeNode root, int targetSum) {
        List<Integer> path = new ArrayList<>();
        if(root==null)return res;
        path.add(root.val);
        helper(root,root.val,targetSum,path);
        return res;
    }
}